### PR TITLE
Add a proof checklist to the training material.

### PR DESCRIPTION
**Pull request by markrtuttle**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: training-material/checklist.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  The starter kit uses these flags by default.  But the user of the starter kit
+  may disable any one of these flags by editing project Makefile.common or
+  by setting a makefile variable to the empty string
+  (as in `CBMC_FLAG_MALLOC_MAY_FAIL = `)
+  in the project Makefile-project-defines or a proof Makefile.
+
+* Document the decision to omit one of these property-checking flags.
+  There are valid reasons to omit them either for a project or for an
+  individual proof. But the decision and the reason for the decision
+  must be document either in a project readme or a proof readme file.
```
Comment (nchong-at-aws) [inline training-material/checklist.md:54]: *documented

File: training-material/checklist.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+* Define an
+  [`is_valid()` predicate](PROOF-WRITING.md##the-is_valid-function)
+  as describe in the training material for every nontrivial data structure.
+
+* Gather the definitions of `ensure_allocated` and `is_valid` in a common
+  location, most commonly in the `proofs/sources` subdirectory of the
+  starter kit.
+
+* Check that every instance of `__CPROVER_assume` appears either a proof
```
Comment (nchong-at-aws) [inline training-material/checklist.md:76]: appears in?

File: training-material/checklist.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  harness. Some exceptions are required.  For example, it may be necessary
+  in an `ensure_allocated` to assume `length < CBMC_MAX_OBJECT_SIZE` before
+  invoking `malloc(length)` to avoid a false positive about malloc'ing a
+  too-big object. But every instance of `__CPROVER_assume` in supporting code
+  could be copied into the proof harness.  The goal is for all proof
```
Comment (nchong-at-aws) [inline training-material/checklist.md:81]: *could -> should?

File: training-material/checklist.md (hunk 1 above)
Comment (markrtuttle) [reply training-material/checklist.md:54]: Done. Thanks.

File: training-material/checklist.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+* Define an
+  [`ensure_allocated` function](PROOF-WRITING.md#the-ensure_allocated-function)
+  as described in the training material for every nontrivial data structure.
+  Feel free to use any naming scheme that makes sense for your project --- some
+  projects use `allocate_X` in place of `ensure_allocated_X` --- but be
+  consistent.
+
+* Define an
+  [`is_valid()` predicate](PROOF-WRITING.md##the-is_valid-function)
+  as describe in the training material for every nontrivial data structure.
```
[the remark this answers is not in the document]
Comment (markrtuttle) [reply training-material/checklist.md:70]: Done. Thanks.

File: training-material/checklist.md (hunk 2 above)
Comment (markrtuttle) [reply training-material/checklist.md:76]: Done. Thanks.

File: training-material/checklist.md (hunk 3 above)
Comment (markrtuttle) [reply training-material/checklist.md:81]: Done. Thanks.

File: training-material/checklist.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+These answers should be found in one of three places:
+
+* The proof harness,
+* The proof makefiles (and, with the starter kit, these makefiles are
+  the proof Makefile, the project Makefile-project-defines,
+  and the project Makefile.common), and perhaps
+* The proof readme file.
+
+## Properties checked
+
```
Comment (mww-aws) [inline training-material/checklist.md:28]: It is a stylistic thing, but the current doc reads as best practices, not a checklist.  For example: 

* Document the decision to omit one of these property-checking flags.
  There are valid reasons to omit them either for a project or for an
  individual proof. But the decision and the reason for the decision
  must be documented either in a project readme or a proof readme file.

How about instead we have: 
* All deviations from the standard property-checking flags are documented.

  There are valid reasons to omit them either for a project or for an
  individual proof. But the decision and the reason for the decision
  must be documented either in a project readme or a proof readme file.

The latter form is something that can easily be "checked off".

File: training-material/checklist.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+* Document the decision to omit one of these property-checking flags.
+  There are valid reasons to omit them either for a project or for an
+  individual proof. But the decision and the reason for the decision
+  must be documented either in a project readme or a proof readme file.
+
+* CBMC checks assertions in the code.  This is understood and need not be
+  documented.
+
+## Assumptions made
+
```
Comment (mww-aws) [inline training-material/checklist.md:60]: Another example: 
* For each non-trivial data structure, an [`ensure_allocated` function](PROOF-WRITING.md#the-ensure_allocated-function) as described in the training material, has been defined.

  Feel free to use any naming scheme that makes sense for your project --- some
  projects use `allocate_X` in place of `ensure_allocated_X` --- but be
  consistent.

**Review by mww-aws — CHANGES_REQUESTED** [review]

File: training-material/checklist.md (hunk 6 above)
Comment (markrtuttle) [reply training-material/checklist.md:65]: Done.  Thanks.

**Review by mww-aws — APPROVED** [review]
Thanks Mark.  Looks good.

**Review by danielsn — APPROVED** [review]

**Pull request merged**